### PR TITLE
Consolidate VS detection logic

### LINKED ISSUE #1251: Centralize and improve Windows prereq checks
I got the missing DIA SDK error when first building on my laptop just now:
```
Error: DIA SDK is missing at "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\DIA SDK". This is due to a bug in the Visual Studio installer. It does not install DIA SDK at "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\" but rather at the install location of previous Visual Studio version. The workaround is to copy the DIA SDK folder from the Visual Studio install location of the previous version to "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\" and then build.
Visual Studio Express does not include the DIA SDK. You need Visual Studio 2017 or 2019 (Community is free).
```

In the current machine, I have 15.9.11 Preview 1 and 16.5.0 Preview 2. I am building in the VS 2019 prompt, where `VSINSTALLDIR=C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\`. However the single DIA SDK folder is under the VS 2017 prompt's `VSINSTALLDIR=C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise`. 

I copied it from there to solve the problem, so the message may be fully accurate. However, I have seen this prompt at least 3 years ago, so it is not clear that this will solve itself. Can we look in more places for it, to remove this stumbling block when building the repo the first time?

Perhaps it could look under `C:\Program Files (x86)\Microsoft Visual Studio` for `DIA SDK` if it can't find one under VSINSTALLDIR?

### PR DESCRIPTION
**Pull request by AntonLapounov**
Consolidate and improve Visual Studio installation detection logic. Fixes #1251.

At present we use four almost identical code snippets in different script files for detection Visual Studio installation. They all run `vswhere.exe`, check Visual Studio version, launch Developer Command Prompt, and optionally set VC++ build environment, check DIA SDK's presence, and set `CMakePath` environment variable. The goal of this PR is to consolidate that logic in a single script to improve maintainability.

The current VS detection logic has the following flaw. If you have several VS installations on the machine, it may choose a VS installation with no VC Tools component present, which will fail any native build. Since DIA SDK is always installed with the VC Tools component, checking for DIA SDK's presence in our scripts is more or less equivalent to checking for the VC Tools component's presence. This PR fixes the described flaw by choosing only among VS installations with the VC Tools component present (still choosing the latest VS if there are many).

One consequence of this PR will be failing the build earlier if there is no VS installation with the VC Tools component present. If we think purely managed builds with no VC Tools installed on the computer is a scenario worth supporting, I can relax the `VSWhere`'s filter to require the VC Tools component only when the native architecture is passed to `init-vs-env.cmd`.

Below I summarized the relevant steps different scripts would execute before and after my changes.
Before | After
-- | --
**eng\native\init-vs-env.cmd**<br/>**This script did not exist**<br/>&nbsp;<br/>&nbsp;<br/>&nbsp; | • Runs `VSWhere` and calls `VsDevCmd.bat -no_logo`<br/>• VC Tools component check guarantees DIA SDK's presence<br/>_Only if architecture is passed as an argument:_<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Sets `CMakePath`
**src\coreclr\build-runtime.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat` (via `setup_vs_tools.cmd`)<br/>_Only if native build is needed:_<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Checks DIA SDK's presence<br/>• Sets `CMakePath` | Same list but calls `init-vs-env.cmd` instead of `setup_vs_tools.cmd`<br/>in the first step (checking for VC Tools _unconditionally_ instead of<br/>the conditional DIA SDK check)
**src\coreclr\crossgen-corelib.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat` (via `setup_vs_tools.cmd`)<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Checks DIA SDK's presence | Calls `init-vs-env.cmd %__BuildArch%`, which does all the steps on the left (checking for VC Tools instead of DIA SDK)
**src\coreclr\crossgen-corelib.proj**<br/>• Runs `VSWhere `and calls `VsDevCmd.bat` (via `setup_vs_tools.cmd`) | Same but calls `init-vs-env.cmd` instead of `setup_vs_tools.cmd`
**src\coreclr\setup_vs_tools.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat` | **This script is replaced by `init-vs-env.cmd`**
**src\libraries\Native\build-native.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat -no_logo`<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Checks DIA SDK's presence<br/>• Sets `CMakePath` | Calls `init-vs-env.cmd %__BuildArch%`, which does all the steps on the left (checking for VC Tools instead of DIA SDK)
**src\mono\mono.proj**<br/>• Calls `init-compiler-and-cmake.cmd $(Platform)` (see the next line) | Calls `init-vs-env.cmd $(Platform)`
**eng\native\init-compiler-and-cmake.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat -no_logo`<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Checks DIA SDK's presence<br/>• Sets `CMakePath` | **This script is replaced by `init-vs-env.cmd`**
**src\native\corehost\build.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat`<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Checks DIA SDK's presence<br/>• Sets `CMakePath` | Calls `init-vs-env.cmd %__BuildArch%`, which does all the steps on the left (checking for VC Tools instead of DIA SDK)
**src\tests\build.cmd**<br/>• Runs `VSWhere` and calls `VsDevCmd.bat` (via `setup_vs_tools.cmd`)<br/>_Only if native build is needed:_<br/>• Calls `vcvarsall.bat %__VCBuildArch%`<br/>• Checks DIA SDK's presence<br/>• Sets `CMakePath` | Same list but calls `init-vs-env.cmd` instead of `setup_vs_tools.cmd`<br/>in the first step (checking for VC Tools _unconditionally_ instead of<br/>the conditional DIA SDK check)

### CODE REVIEW

**Comment by AntonLapounov** [comment]
The Mono failure is #49569.

**Comment by AntonLapounov** [comment]
Cc @trylek regarding the plan he suggested in https://github.com/dotnet/runtime/issues/38919#issuecomment-661317709.

File: eng/native/init-vs-env.cmd [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+set "__VSCOMNTOOLS="
+set "VSCMD_START_DIR="
+
+:VSDetected
+if "%VisualStudioVersion%"=="16.0" (
+    set __VSVersion=vs2019
+    set __PlatformToolset=v142
+    goto :SetVCEnvironment
+) else if "%VisualStudioVersion%"=="15.0" (
+    set __VSVersion=vs2017
```
Comment (jkotas) [inline eng/native/init-vs-env.cmd:53]: We do not support building with VS2017 anymore. You can delete this part.

**Comment by jkotas** [comment]
> Contributes to #1251.

Is there anything else to do to fix #1251? I think that this is complete fix.

**Comment by AntonLapounov** [comment]
Yes, I hope we can close #1251 with this fix. I was going to check that with @danmoseley.

**Comment by danmoseley** [comment]
Sounds like a fix to me. Thanks for the cleanup!

File: eng/native/init-vs-env.cmd (hunk 1 above)
Comment (AntonLapounov) [reply eng/native/init-vs-env.cmd]: Also removed two remaining DIA SDK checks in the two scripts that call `init-vs-env.cmd` without passing an architecture and later optionally run `vcvarsall.bat %__VCBuildArch%`. With the current changes, successful execution of `init-vs-env.cmd` implies we must have DIA SDK. If we decide to relax that requirement for building managed components only, we will have to put DIA SDK checks or equivalent VC Tools checks back.

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by ViktorHofer** [comment]
> One consequence of this PR will be failing the build earlier if there is no VS installation with the VC Tools component present. If we think purely managed builds with no VC Tools installed on the computer is a scenario worth supporting, I can relax the VSWhere's filter to require the VC Tools component only when the native architecture is passed to init-vs-env.cmd.

Just to double check, I don't  see a reference to the consolidated script from the base build.ps1/build.sh scripts which makes me believe that managed builds don't fail early, right? I would expect to be able to i.e. build the libs.ref subset without a prereq check.

**Comment by ViktorHofer** [comment]
OK just saw that your comment refers to the native check in the build-runtime.cmd and in the src/tests/build.cmd script. I think it's fine to unconditionally check for the VS installation.

**Review by ViktorHofer — APPROVED** [review]

**Review by trylek — APPROVED** [review]
Awesome cleanup, thank you!

**Comment by akoeplinger** [comment]
Thank you for this cleanup!

**Pull request merged**